cunumeric/vectorize.py

```diff
@@ -0,0 +1,643 @@
+                    " for each invocation of the vectorize object"
+                )
+        elif self._cache:
+            self._cached_dtype = output_dtype
+
+        # FIXME
+        # we could find common type of input arguments here and
+        # broadcasted shapes
+        if self._num_outputs > 0 and output_dtype is None:
+            raise ValueError("Unable to choose output dtype")
+        if self._num_outputs > 0 and output_shape is None:
+            raise ValueError("Unable to choose output shape")
+
+        # filing the list of return arguments
+        # check if there are return argnames in input argnames,
+        # if not, create a new array
+        for r in self._return_names:
+            if r in self._arg_names:
+                idx = self._arg_names.index(r)
+                if self._args[idx].shape != output_shape:
+                    raise ValueError(
+                        "all output arrays should have the same shape"
+                    )
+                if output_dtype != self._args[idx].dtype:
+                    runtime.warn(
+                        "converting input array to output types in user func ",
+                        category=RuntimeWarning,
+                    )
+                    self._args[idx] = self._args[idx].astype(output_dtype)
+                self._return_args.append(self._args[idx])
+                self._args.remove(self._args[idx])
+                self._arg_names.remove(r)
+            elif r in self._scalar_names:
+                idx = self._scalar_names.index(r)
+                if output_shape != (1,):
+                    raise ValueError(
+                        "all output arrays should have the same shape"
+                    )
+                self._return_args.append(
+                    full(output_shape, self._scalar_args[idx], output_dtype)
+                )
+                self._scalar_args.remove(self._scalar_args[idx])
+                self._scalar_names.remove(r)
+            else:
+                # create array and add it to the list of return_args
+                tmp_ret = full(output_shape, 0, output_dtype)
+                self._return_args.append(tmp_ret)
+
+        # check types and shapes
+        if len(self._args) > 0:
+            for count, a in enumerate(self._args):
+                if output_dtype != a.dtype:
+                    runtime.warn(
+                        "converting input array to output types in user func ",
+                        category=RuntimeWarning,
+                    )
+                    self._args[count] = self._args[count].astype(output_dtype)
+                # FIXME broadcast shapes
+                if output_shape != self._args[count].shape:
+                    raise ValueError(
+                        "cuNumeric doesn't support "
+                        "different shapes for arrays in "
+                        "user function passed to vectorize"
+                    )
```

Comment:
This method is quite long, maybe some of the checks and computations could be factored into helper methods?